internal/store/device.go

```diff
@@ -182,15 +243,77 @@ func (s *DeviceStore) UpdateStatus(ctx context.Context, orgId uuid.UUID, resourc
 	return resource, result.Error
 }
 
-func (s *DeviceStore) Delete(ctx context.Context, orgId uuid.UUID, name string) error {
-	condition := model.Device{
-		Resource: model.Resource{OrgID: orgId, Name: name},
+func (s *DeviceStore) Delete(ctx context.Context, orgId uuid.UUID, name string, callback DeviceStoreCallback) error {
+	var existingRecord model.Device
+	err := s.db.Transaction(func(innerTx *gorm.DB) (err error) {
+		existingRecord = model.Device{Resource: model.Resource{OrgID: orgId, Name: name}}
+		result := innerTx.First(&existingRecord)
+		if result.Error != nil {
+			return result.Error
+		}
+		result = innerTx.Unscoped().Delete(&existingRecord)
+		return result.Error
+	})
+
+	if err == nil {
+		callback(&existingRecord, nil)
 	}
-	result := s.db.Unscoped().Delete(&condition)
-	return result.Error
+	if errors.Is(err, gorm.ErrRecordNotFound) {
+		return nil
+	}
+
+	return err
+}
+
+func (s *DeviceStore) UpdateOwner(ctx context.Context, orgId uuid.UUID, name string, owner string, callback DeviceStoreCallback) error {
+	device := model.Device{Resource: model.Resource{OrgID: orgId, Name: name}}
+	err := s.db.Model(device).Updates(map[string]interface{}{
+		"generation": 0,
+		"owner":      owner,
+	}).Error
+	if err == nil {
+		before := &device
+		after := &model.Device{
+			Resource: model.Resource{
+				OrgID:      orgId,
+				Name:       name,
+				Owner:      &owner,
+				Generation: util.Int64ToPtr(0),
+			}}
+		callback(before, after)
+	}
+	return err
+}
+
+func (s *DeviceStore) UpdateAnnotations(ctx context.Context, orgId uuid.UUID, name string, annotations *map[string]string, deleteKeys *[]string) error {
```

Comment:
nit: it's strange to have a param that is a reference to a map. Under the hood it's already a referenced type.  same with `LabelMapToArray` if you want to know if its empty just check len.

```
if len(annotations) == 0 {
  // do something
}
```

Reply:
Same for slices.

Reply:
Fixed